x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java

```diff
@@ -347,20 +374,33 @@ public void onFailure(Exception e) {
 
                 @Override
                 protected void doRun() throws Exception {
-                    final BulkShardOperationsRequest request = new BulkShardOperationsRequest(followerShard, response.getOperations());
-                    followerClient.execute(BulkShardOperationsAction.INSTANCE, request, new ActionListener<BulkShardOperationsResponse>() {
-                        @Override
-                        public void onResponse(final BulkShardOperationsResponse bulkShardOperationsResponse) {
-                            handler.accept(null);
-                        }
-
-                        @Override
-                        public void onFailure(final Exception e) {
-                            // No retry mechanism here, because if a failure is being redirected to this place it is considered
-                            // non recoverable.
-                            assert e != null;
-                            handler.accept(e);
+                    indexVersionChecker.accept(response.getIndexMetadataVersion(), e -> {
+                        if (e != null) {
+                            if (retryCounter.incrementAndGet() <= PROCESSOR_RETRY_LIMIT) {
```

Comment:
do we want a retry on every exception type?

Reply:
Good question, I think not.

Reply:
I made this change: b569f88d3aabea86fd7bf2093f25530005669cfb